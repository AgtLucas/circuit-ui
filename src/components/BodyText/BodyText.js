import React from 'react';
import PropTypes from 'prop-types';
import { css } from 'react-emotion';

import { standard } from '../../themes';
import { typeMarginResets } from '../../styles/global-styles';

const baseStyles = ({ theme }) => css`
  label: body-text;
  font-weight: ${theme.fontWeight.regular};
  ${typeMarginResets};
`;

const sizeStyles = ({ theme, size }) => css`
  label: body-text--${size};
  font-size: ${theme.typography.body[size].fontSize};
  line-height: ${theme.typography.body[size].lineHeight};
`;

/**
 * The BodyText component is used for long-form text. Typically with
 * <p>, <div>, <article>, or <section> elements.
 */
const BodyText = ({
  element: Element,
  size,
  theme,
  children,
  className,
  ...props
}) => (
  <Element
    className={`${baseStyles({ theme })} ${sizeStyles({
      theme,
      size
    })} ${className}`}
    {...props}
  >
    {children}
  </Element>
);

BodyText.propTypes = {
  /**
   * Child nodes to be rendered.
   */
  children: PropTypes.oneOfType([
    PropTypes.arrayOf(PropTypes.node),
    PropTypes.node
  ]),
  /**
   * A Circuit UI body text size.
   */
  size: PropTypes.oneOf(['kilo', 'mega', 'giga']),
  /**
   * Optional additional className string to overwrite styles.
   */
  className: PropTypes.string,
  /**
   * The HTML element to render.
   */
  element: PropTypes.string,
  /**
   * A Circuit UI theme object. Usually provided by a ThemeProvider.
   */
  theme: PropTypes.object
};

BodyText.defaultProps = {
  element: 'p',
  size: 'mega',
  className: '',
  theme: standard
};

export default BodyText;
