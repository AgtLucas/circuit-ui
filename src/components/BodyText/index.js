import { withTheme } from 'emotion-theming';

import PlainBodyText from './BodyText';

export { PlainBodyText };

export default withTheme(PlainBodyText);
