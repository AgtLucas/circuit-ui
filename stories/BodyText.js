import React from 'react';
import { flow } from 'lodash/fp';
import { storiesOf } from '@storybook/react';
import { withNotes } from '@storybook/addon-notes';
import { withInfo } from '@storybook/addon-info';

import withTests from '../src/util/withTests';
import { PlainBodyText as BodyText } from '../src/components/BodyText';
import README from '../src/components/BodyText/README.md';

storiesOf('BodyText', module)
  .addDecorator(withTests('BodyText'))
  .add(
    'BodyTexts',
    flow(withNotes(README), withInfo())(() => [
      <BodyText key="pkilo" element="p" size="giga">
        This is a giga BodyText with a p element
      </BodyText>,
      <BodyText key="articlemega" element="article" size="mega">
        This is an mega BodyText with an article element
      </BodyText>,
      <BodyText key="divkilo" element="div" size="kilo">
        This is a kilo BodyText with a div element
      </BodyText>
    ])
  );
