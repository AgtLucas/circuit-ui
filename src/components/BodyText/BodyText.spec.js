import React from 'react';
import renderer from 'react-test-renderer';

import BodyText from './BodyText';

describe('BodyText', () => {
  const elements = ['p', 'article', 'div'];
  elements.forEach(el => {
    it(`should render as ${el} element, when passed "${el}" for the element prop`, () => {
      const heading = renderer
        .create(
          <BodyText element={el}>{`${el.toUpperCase()} heading`}</BodyText>
        )
        .toJSON();
      expect(heading).toMatchSnapshot();
    });
  });
  const sizes = ['kilo', 'mega', 'giga'];
  sizes.forEach(size => {
    it(`should render with size ${size}, when passed "${size}" for the size prop`, () => {
      const heading = renderer
        .create(<BodyText {...{ size }}>{`${size} heading`}</BodyText>)
        .toJSON();
      expect(heading).toMatchSnapshot();
    });
  });
});
