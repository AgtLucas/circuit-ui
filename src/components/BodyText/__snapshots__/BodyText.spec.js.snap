// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BodyText should render as article element, when passed "article" for the element prop 1`] = `
.circuit-0 {
  font-weight: 400;
  -webkit-margin-before: 0;
  -webkit-margin-after: 0;
  -webkit-margin-start: 0;
  -webkit-margin-end: 0;
}

.circuit-1 {
  font-size: 15px;
  line-height: 24px;
}

<article
  className="circuit-0 circuit-1 "
>
  ARTICLE heading
</article>
`;

exports[`BodyText should render as div element, when passed "div" for the element prop 1`] = `
.circuit-0 {
  font-weight: 400;
  -webkit-margin-before: 0;
  -webkit-margin-after: 0;
  -webkit-margin-start: 0;
  -webkit-margin-end: 0;
}

.circuit-1 {
  font-size: 15px;
  line-height: 24px;
}

<div
  className="circuit-0 circuit-1 "
>
  DIV heading
</div>
`;

exports[`BodyText should render as p element, when passed "p" for the element prop 1`] = `
.circuit-0 {
  font-weight: 400;
  -webkit-margin-before: 0;
  -webkit-margin-after: 0;
  -webkit-margin-start: 0;
  -webkit-margin-end: 0;
}

.circuit-1 {
  font-size: 15px;
  line-height: 24px;
}

<p
  className="circuit-0 circuit-1 "
>
  P heading
</p>
`;

exports[`BodyText should render with size giga, when passed "giga" for the size prop 1`] = `
.circuit-0 {
  font-weight: 400;
  -webkit-margin-before: 0;
  -webkit-margin-after: 0;
  -webkit-margin-start: 0;
  -webkit-margin-end: 0;
}

.circuit-1 {
  font-size: 18px;
  line-height: 28px;
}

<p
  className="circuit-0 circuit-1 "
>
  giga heading
</p>
`;

exports[`BodyText should render with size kilo, when passed "kilo" for the size prop 1`] = `
.circuit-0 {
  font-weight: 400;
  -webkit-margin-before: 0;
  -webkit-margin-after: 0;
  -webkit-margin-start: 0;
  -webkit-margin-end: 0;
}

.circuit-1 {
  font-size: 13px;
  line-height: 20px;
}

<p
  className="circuit-0 circuit-1 "
>
  kilo heading
</p>
`;

exports[`BodyText should render with size mega, when passed "mega" for the size prop 1`] = `
.circuit-0 {
  font-weight: 400;
  -webkit-margin-before: 0;
  -webkit-margin-after: 0;
  -webkit-margin-start: 0;
  -webkit-margin-end: 0;
}

.circuit-1 {
  font-size: 15px;
  line-height: 24px;
}

<p
  className="circuit-0 circuit-1 "
>
  mega heading
</p>
`;
